Tk regex

ocap-kirk opened the pull request:
## Key Features Added
1. Support for regex pattern matching in URL paths
2. Ability to capture named groups from regex patterns as attributes
3. Fallback to numbered capture groups if named groups aren't present
4. Error handling for invalid regex patterns
5. Logging of regex match results for debugging

## Step-by-Step Changes

1. **Schema Updates**:
   - Added `url_type` field to `MappingRuleData` to distinguish between regex and traditional patterns
   - Updated validation to support regex patterns

2. **URL Comparison**:
   - Modified `_compare_urls` to handle both traditional and regex patterns
   - Added `is_regex` flag to control matching behavior
   - Implemented regex pattern compilation and matching

3. **Attribute Extraction**:
   - Created separate paths for regex vs traditional attribute extraction
   - Added support for named capture groups in regex patterns
   - Implemented fallback to numbered capture groups
   - Maintained backward compatibility with existing URL parameter extraction

4. **Error Handling**:
   - Added try-catch blocks for regex compilation
   - Implemented logging for regex-related errors
   - Added fallback behavior for invalid patterns

5. **Logging & Debugging**:
   - Added detailed logging for regex matching results
   - Included pattern and URL information in logs
   - Added debug information for troubleshooting

danyi1212 commented on horizon/enforcer/schemas.py:
```diff
@@ -141,12 +141,13 @@ class AllTenantsAuthorizationResult(BaseSchema):
     resource: str
     action: str
     priority: int | None = None
+    url_type: Optional[str] = None
```
Better to define as a string enum
```python
class UrlTypes(StrEnum):
    regex = "regex"
    ...
```

ocap-kirk replied:
Agreed.

ocap-kirk replied:
Also, should we add that the PDP is now dependant on Python v3.11+? `StrEnum` was introduced in 3.11. I can make it 3.10 compliant by `class UrlTypes(str, Enum):`

ocap-kirk replied:
we may already be 3.11+ with the PDP, but it isn't in the README

danyi1212 replied:
Correct, the PDP is still Python v3.10. We can upgrade, but let's not involve too many things, use the `class UrlTypes(str, Enum):` for now

ocap-kirk replied:
Fixed.

danyi1212 commented on requirements.txt:
```diff
@@ -14,3 +14,4 @@ httpx>=0.27.0,<1
 protobuf>=3.20.2 # not directly required, pinned by Snyk to avoid a vulnerability
 opal-common==0.7.15
 opal-client==0.7.15
+google-re2
```
What is the reasoning for using Google's regex instead of the standard library?\

ocap-kirk replied:
Many reasons: RE2 is built for untrusted input eliminates backtracking. It is essentially immune to Regular Expression Denial of Service (ReDoS) attacks. It doesn't use recursion. Battle tested for production. It also has predictable performance, guaranteeing O(n).

I mainly want to avoid the risks of using python re library which handles much more than what we need in this feature and adds too much risk.

danyi1212 commented on horizon/enforcer/utils/mapping_rules_utils.py:
```diff
@@ -102,25 +104,84 @@ def extract_attributes_from_query_params(rule_url: str, request_url: str) -> dic
         for mapping_rule in mapping_rules:
-            if not mapping_rule.http_method == http_method.lower():
-                # if the method is not the same, we don't need to check the url
+            is_regex = getattr(mapping_rule, 'url_type', None) == "regex"
```
According to the type annotation, this should be a pydantic instance, meaning it must have `url_type`. Why did you use `getattr()` instead?
 ```suggestion
            is_regex = mapping_tule.url_type == "regex"
```

ocap-kirk replied:
Good catch. Fixed.

danyi1212 commented on horizon/enforcer/utils/mapping_rules_utils.py:
```diff
@@ -102,25 +104,84 @@ def extract_attributes_from_query_params(rule_url: str, request_url: str) -> dic
-            if not cls._compare_urls(mapping_rule.url, url):
-                # if the urls doesn't match, we don't need to check the headers
+                
+            if not cls._compare_urls(mapping_rule.url, url, is_regex=is_regex):
```
Why did you remove the inline comments?

ocap-kirk commented on horizon/enforcer/schemas.py:
```diff
@@ -141,12 +146,13 @@ class AllTenantsAuthorizationResult(BaseSchema):
     resource: str
     action: str
     priority: int | None = None
+    url_type: Optional[UrlTypes] = None
```
added default var.

ocap-kirk commented:
There was a bug with existing {} variables that I fixed and I added a few more tests to test the existing functionality. I also tested E2E and everything looks good.

omer9564 commented on horizon/enforcer/utils/mapping_rules_utils.py:
```diff
@@ -88,25 +91,81 @@ def extract_attributes_from_query_params(rule_url: str, request_url: str) -> dic
                 # if the urls doesn't match, we don't need to check the headers
                 continue
-            matched_mapping_rules.append(mapping_rule)
-        # most priority first
-        matched_mapping_rules.sort(key=lambda rule: rule.priority or 0, reverse=True)
-        if len(matched_mapping_rules) > 0:
-            return matched_mapping_rules[0]
```
why remove this ?

omer9564 commented on horizon/enforcer/api.py:
```diff
@@ -748,3 +762,66 @@ async def is_allowed_kong(request: Request, query: KongAuthorizationQuery):
+        return {}
+
+
+def _extract_url_attributes(matched_rule: MappingRuleData, url: str) -> dict:
```
this is unused ?
